RELEASE/scripts/autoscend.ash

```diff
@@ -2384,20 +2384,140 @@ boolean powerLevelAdjustment()
 	return false;
 }
 
+boolean unlockThinknerdWarehouse()
+{
+	//returns true if you have access to [The Thinknerd Warehouse], if you don't have access tries to unlock it and then return true.
+	//returns false if fails to unlock.
+	//much easier to do if you already have torso awaregness
+	
+	if(internalQuestStatus("questM22Shirt") > -1)
+	{
+		return true;
+	}
+	
+	//unlocking is a multi step process. We want to try things in reverse to conserve resources and in case some steps were already complete.
+	
+	boolean useLetter()
+	{
+		if(item_amount($item[Letter for Melvign the Gnome]) > 0)
+		{
+			if(use(1, $item[Letter for Melvign the Gnome]))
+			{
+				return true;
+			}
+			else
+			{
+				abort("Somehow failed to use [Letter for Melvign the Gnome]... aborting to prevent infinite loops");
+			}
+		}
+		return false;
+	}
+	
+	item target_shirt = $item[none];		//needs to be global so we can call it outside boolean hasShirt()
+	boolean hasShirt()
+	{
+		if(target_shirt != $item[none]) return true;
+		int[item] inventory_snapshot = get_inventory();		//need to refresh inventory_snapshot every time this function is called.
+		foreach it in inventory_snapshot
+		{
+			if(item_type(it) == "shirt" && can_equip(it))
+			{
+				target_shirt = it;
+				break;
+			}
+		}
+		if(target_shirt != $item[none]) return true;
+		return false;
+	}
+	boolean useShirtThenLetter()
+	{
+		equip($slot[shirt], target_shirt);
+		if(useLetter()) return true;
+		auto_log_error("For some reason boolean unlockThinknerdWarehouse() failed when trying to use the shirt [" + target_shirt + "] to get [Letter for Melvign the Gnome] to start the quest", "red");
+		return false;
+	}
+	void tryShirtCreate(item it)
+	{
+		if(!hasShirt() && can_equip(it) && creatable_amount(it) > 0)
+		{
+			if(create(1, it))
+			{
+				target_shirt = it;
+			}
+		}
+	}
+	void tryShirtPull(item it)
+	{
+		if(!hasShirt() && can_equip(it) && canPull(it))
+		{
+			if(pullXWhenHaveY(it, 1, 0))
+			{
+				target_shirt = it;
+			}
+		}
+	}
+	
+	//if you already had a shirt or a letter, then just unlock the quest now
+	if(useLetter()) return true;
+	if(hasShirt())
+	{
+		if(useShirtThenLetter()) return true;
+	}
+	
+	//Try to acquire a shirt.
+	//IOTM foldable shirts that cost nothing.
+	//TODO, make these actually work
+	//	tryShirtCreate($item[flaming pink shirt])		//foldable IOTM
+	//	tryShirtCreate($item[origami pasties])			//foldable IOTM
+	// 	tryShirtCreate($item[sugar shirt])				//libram summons sugar sheet which can be folded into sugar shirt
+	//	tryShirtCreate($item[makeshift garbage shirt])	//actually needs januaryToteAcquire($item[makeshift garbage shirt])
+	
+	//Spend a pull on a shirt
+	tryShirtPull($item[Sneaky Pete\'s leather jacket]);		//useful IOTM shirt with no requirements to wear
```

Comment:
could this not be

```
boolean hasShirt;

boolean pullWhenHaveNoShirt(item it) {
  if (hasShirt) return true;
  return can_equip(it) && pullXWhenHaveY(it, 1, 0));
}

boolean = pullWhenHaveNoShirt($item[foo]);
boolean = pullWhenHaveNoShirt($item[bar]);
```

etc.

Partly this is a naming objection - when I first looked at this code it looked like it would pull all the listed shirts. Plus, `hasShirt` doesn't need to be a function that does a linear scan through your entire inventory for every shirt you consider obtaining.

Reply:
The first line in hasShirt() is
`if(target_shirt != $item[none]) return true;`
tryShirtPull and tryShirtCreate both sets the variable `target_shirt` to something that is not none if they succeed in creating or pulling a shirt.
Which means the moment you succeed on one of those then all future attempts will have hasShirt return true on the first line, before it actually calls on your full inventory. And will thus skip the pull or create attempt

Reply:
For the sake of code clarity. I could add a redundant
`boolean has_shirt = false;` at the begining.
Currently when successfully pull/craft a shirt it changes `target_shirt` from none to the name of the shirt. I could have it also change has_shirt bool to true at that point.

Would add 3 unnecessary lines of code. But potentially be a lot better at self documenting?

Reply:
facepalm... i just realized that this line of mine only stops it from running through entire inventory IF you have a shirt already... so long as you are failing it does repeat the check.
So those 3 lines of code would not be redundant at all...

Yea, I will redo this tomorrow. my bad there

Reply:
change made